src/geometries/primitives/conesurface.jl

```diff
@@ -31,17 +31,19 @@ apex(c::ConeSurface) = c.apex
 Base.isapprox(c₁::ConeSurface, c₂::ConeSurface; atol=atol(lentype(c₁)), kwargs...) =
   isapprox(c₁.base, c₂.base; atol, kwargs...) && isapprox(c₁.apex, c₂.apex; atol, kwargs...)
 
-function (c::ConeSurface)(φ, h)
-  T = numtype(lentype(c))
-  if (φ < 0 || φ > 1) || (h < 0 || h > 1)
-    throw(DomainError((φ, h), "c(φ, h) is not defined for φ, h outside [0, 1]²."))
+function (conesurface::ConeSurface)(uφ, uh)
+  T = numtype(lentype(conesurface))
+  if (uφ < 0 || uφ > 1) || (uh < 0 || uh > 1)
+    throw(DomainError((uφ, uh), "c(φ, h) is not defined for φ, h outside [0, 1]²."))
   end
-  n = -normal(c.base)
-  v = c.base(T(0), T(0)) - c.apex
-  l = norm(v)
-  θ = ∠(n, v)
-  o = c.apex + T(h) * v
-  r = T(h) * l * cos(θ)
-  s = Circle(Plane(o, n), r)
-  s(T(φ))
+
+  a = conesurface.apex
+  R = conesurface.base.radius
+  b = conesurface.base.plane.p
+  û = normalize(conesurface.base.plane.u)
+  v̂ = normalize(conesurface.base.plane.v)
+  sφ, cφ = sincospi(2uφ)
+  c = b + Vec(R * cφ * û) + Vec(R * sφ * v̂)
+  h̄ = uh * (c - a)
+  a + h̄
```

Comment:
Please use the public api of the geometries instead of accessing fields directly as they may change in the future. Given that this is a function of the `ConeSurface`, you can do `c.base`, but `c.base.radius` is accessing an implementation detail of the `base`. You can write `radius(c.base)` instead.

Reply:
Here you can use the parametrization of the `Disk` as the public api to generate points.

Reply:
So, long-term I really want to get AD rules defined for all of these parametric functions. From what I'm reading and a little bit of experimentation, it looks like we're going to have to define explicit AD rules for all of these functions to get that working.

One of my goals with the approach I used here was to keep the math concrete and simple to code up analytic derivation rules for. By contrast, if we rely on abstract API's and don't control for lower-level implementation details, then we'd need to propagate a more complicated chain-rules analysis (beyond my AD skills here).

Reply:
> So, long-term I really want to get AD rules defined for all of these parametric functions

Very nice 👍🏽 

> From what I'm reading and a little bit of experimentation, it looks like we're going to have to define explicit AD rules for all of these functions to get that working.

Can you please clarify how using the public API interferes with AD rules? The function calls should produce the exact same inline code if I understand it correctly.

Reply:
I'm not an AD SME, so if a simpler solution exists I'm all for it. I've yet to get any AD package to calculate derivatives for any of these parametric functions out-of-the-box, even for a simple one like `(s::Segment)(t)` whose derivative should just be `(b - a)::Vec`.

It seems like we need to define `ChainRulesCore.rrule`s for these functions with analytic solutions. Under that scenario, the function above is trivially derived w.r.t. its arguments, but if we have to rely on other parametric functions and treat them as public API, then we need to figure out significantly more complicated rules.

I'm just not sure that the theoretical maintenance burden vs perfect abstractions is worth the added complexity, at least in this particular case.

Reply:
From what I understand, if we express `cylinder(u,v,w)` in terms of `disk(u, v)` plus other expressions, it should be easy to define the chain rules for the disk parametrization and reuse in the cylinder parametrization.

Let's stay consistent with the public API and revisit this issue in the future when you start working with AD definitions. The entire Meshes.jl project is written that way to facilitate fixes and refactorization. If functions in different source files start to use internal fields of structs defined elsewhere, things can get out of control very quickly.